packages/babel/package.json

```
@@ -13,10 +13,14 @@
   "author": "Rich Harris",
   "homepage": "https://github.com/rollup/plugins/tree/master/packages/babel#readme",
   "bugs": "https://github.com/rollup/plugins/issues",
-  "main": "dist/index.js",
-  "module": "dist/index.es.js",
+  "main": "./dist/cjs/index.js",
+  "module": "./dist/es/index.js",
+  "exports": {
+    "require": "./dist/cjs/index.js",
```

In TypeScript with `moduleResolution: node16` the presence of the `package.json#exports` makes it to only look into `exports`. Thus we need to add the `types` condition here too.
```suggestion
    "types": "./types/index.d.ts",
    "require": "./dist/cjs/index.js",
```

---

The additional consideration is also that with `moduleResolution: node16` this won't work any longer:
```ts
import babel from '@rollup/plugin-babel'
```

TS follows the node's suite and it will assume that `babel` here is the namespace object and not the default export. IIRC we already support named export, perhaps we should just drop the `default` export altogether as it only complicates things for the tooling.

We could juggle this and make those types work with the default export under `moduleResolution: node16` but I'm just not sure if it's worth it.

---

That would surprise me. After all, Node 16 very much supports `import babel from`, which is also validated by the tests. Ava 4 no longer transpiles the test files but runs them as native modules, and I used a hack with a symlink `current-package` to validate the exports condition works correctly in those tests.

---

I also verified that `require` returns the default export because it is actually assigned to `module.exports` via a trick in the rollup.config that uses `Object.assign` to put named exports as properties on the default export for maximum Node 16 compatibility. There is an explicit test somewhere that actually does a `createRequire` to verify that.

---

> After all, Node 16 very much supports import babel from

When importing from a CJS package (and `@rollup/plugin-babel`) it gets `module.exports` and not `module.exports.default` though. Perhaps it works right now because of this `footer`:
https://github.com/rollup/plugins/blob/4e85ed78cd2e941107fdf0e8e118e7bee550109d/shared/rollup.config.js#L17

---

Right, I've posted this after reading your first reply - the second one actually confirms my finding. I wasn't aware of this hack~ being used here 😉

---

The code suggestion related to `exports.types` is still relevant though.

---

Definitely and thanks so much for this, I will add this to all other PRs that have types as well.